In [1]:
from tests import *
import math as m

## 1.1) Testing Acummulator SM

In [2]:
test_accumulator_sm()

Test passed!


## 1.2) Testing Binary_addition SM

In [3]:
test_binary_addition_sm()

Test passed!


In [4]:
def test_binary_addition_sm_2():
    res = Binary_Addition().transduce([(1, 0), (0, 1), (1, 1)])
    assert(res == [1, 1, 0, 1])
    print('Test passed!')

In [5]:
test_binary_addition_sm_2()

Test passed!


## 1.3) Testing Reverser SM

In [6]:
test_reverser_sm()

output_seq = [None, None, None, 'bar', ' ', 'foo', None, None, None]
Test passed!


In [7]:
def test_reverser_sm_2():
    res = Reverser().transduce(list('the') + ['end'] + list(range(3)))
    assert(res == [None, None, None, 'e', 'h', 't', None])
    print('Test passed !')

In [8]:
test_reverser_sm_2()

Test passed !


## 1.4) Testing RNN SM

In [9]:
test_rnn()

Test passed!


# 2) MDP

## 2.1) Limited horizons

In [10]:
class MDP:
    def __init__(self, R, T):
        self.R = R
        self.T = T 
    
    def evaluate_policy(self, a, h):
        if h == 0:
            return np.zeros([self.T[a].shape[1], 1])
        return self.R + np.dot(self.T[a], self.evaluate_policy(a, h-1))

In [11]:
R = np.array([[0, 1, 0, 2]]).T

T = {'b': np.array([[0.0, 0.9, 0.1, 0.0],
                    [0.9, 0.1, 0.0, 0.0],
                    [0.0, 0.0, 0.1, 0.9],
                    [0.9, 0.0, 0.0, 0.1]]),
     'c' : np.array([[0.0, 0.1, 0.9, 0.0],
                     [0.9, 0.1, 0.0, 0.0],
                     [0.0, 0.0, 0.1, 0.9],
                     [0.9, 0.0, 0.0, 0.1]])}

robot = MDP(R, T)

In [12]:
robot.T['b']

array([[0. , 0.9, 0.1, 0. ],
       [0.9, 0.1, 0. , 0. ],
       [0. , 0. , 0.1, 0.9],
       [0.9, 0. , 0. , 0.1]])

#### `Which policy is best if you're starting from state 0 with horizon 2`

In [13]:
robot.evaluate_policy('b', 2)

array([[0.9],
       [1.1],
       [1.8],
       [2.2]])

In [14]:
robot.evaluate_policy('c', 2)

array([[0.1],
       [1.1],
       [1.8],
       [2.2]])

Since 0.9 > 0.1, taking action `b` is the better policy.

### `Which policy is best if you're starting from state 0 with horizon 3`

In [15]:
robot.evaluate_policy('b', 3)

array([[1.17],
       [1.92],
       [2.16],
       [3.03]])

In [16]:
robot.evaluate_policy('c', 3)

array([[1.73],
       [1.2 ],
       [2.16],
       [2.31]])

Since 1.17 < 1.73, taking action `c` is the better policy

## 2.2) At a discount

In [17]:
robot.R

array([[0],
       [1],
       [0],
       [2]])

In [18]:
 0.9*robot.T['c']

array([[0.  , 0.09, 0.81, 0.  ],
       [0.81, 0.09, 0.  , 0.  ],
       [0.  , 0.  , 0.09, 0.81],
       [0.81, 0.  , 0.  , 0.09]])

### `Solving a linear equation for value function`

In [19]:
A =  np.eye(4) - 0.9*robot.T['c']
b = robot.R
V = np.linalg.solve(A, b)
V

array([[6.05288295],
       [6.48663207],
       [6.7519581 ],
       [7.58553317]])

# 5) MDP Implementation

## 5.1) Working with MDPs

### `Q value implementation`

In [20]:
test_value()

Test passed!


In [21]:
def test_value_2():
    q = TabularQ([0, 1, 2, 3], ['b', 'c'])
    q.set(2, 'b', -4)
    q.set(2, 'c', 1)
    assert(value(q, 2) == 1)
    print('Test passed!')

In [22]:
test_value_2()

Test passed!


In [23]:
uni_dist = DDist(dict((e, 0.65) for e in ['b', 'c']))
uni_dist.prob('c')

0.65

In [24]:
uni_dist.support()

dict_keys(['b', 'c'])

In [25]:
#q = TabularQ([0, 1, 2, 3], ['b', 'c'])

In [26]:
#t = rv([tiny_transition(0, 'b').prob(elt) for elt in q.states])

In [29]:
#q_old = cv([value(q, s) for s in q.states])

In [ ]:
#np.dot(t, q_old)

array([[0.]])

In [ ]:
#float(1+np.dot(t, q_old))

1.0

## 5.2) Implement Q-Value Iteration

In [ ]:
test_value_iteration()

Test passed!


In [ ]:
def valeur_x():
    x = 3
    while (m.log((3*x +1)/(x + 1), m.e) - x) <= 0.001:
        x = x - 0.01
    print(x)

In [ ]:
valeur_x()

0.52000000000002


## 5.3) Recending-horizon control and online search

In [28]:
test_q_em()

Tests passed!
